pkg/virt-controller/services/template.go

```diff
@@ -149,6 +159,60 @@ func (t *templateService) RenderLaunchManifest(vmi *v1.VirtualMachineInstance) (
 		MountPath: "/var/run/libvirt",
 	})
 
+	if isSriovVmi(vmi) {
+		// libvirt needs this volume to unbind the device from kernel
```

Comment:
Don't we need a TODO here as well? With these mounts, pod can break the host, can't it?

Reply:
@phoracek I may miss something, but I don't think we are able to get rid of these mounts because libvirt assumes access to them to calculate PCI IDs / IOMMU groups / detach devices from hosts. Maybe I am able to mark some of them read-only to reduce the scope, but I believe at least one of them is needed r/w to unbind the device; and /dev/vfio/ is needed to register the device with vfio subsystem.

Reply:
@phoracek tangentially, what is a valid reason for virt-launcher to stay owned by VMI owner? It seems to me this pod is a service pod and should really be isolated from user meddling.

Reply:
@booxter i guess you are right there

Reply:
> @phoracek tangentially, what is a valid reason for virt-launcher to stay owned by VMI owner? It seems to me this pod is a service pod and should really be isolated from user meddling.

Here are some reasons why, just to linke this discussion: https://github.com/kubevirt/kubevirt/pull/1625/files#r227701675

Reply:
@booxter: I have a couple more questions here (some of them may be naive):

> @phoracek I may miss something, but I don't think we are able to get rid of these mounts because libvirt assumes access to them to calculate PCI IDs / IOMMU groups / detach devices from hosts. 

So I guess libvirt has to unbind all related devices (which were made ready for consumption by the device plugin) from the host and then create a group out of it?

> Maybe I am able to mark some of them read-only to reduce the scope, but I believe at least one of them is needed r/w to unbind the device;

Do we have to give the device back if the container goes down dirty? Is unbind here related to unbinding it from the host?

> and /dev/vfio/ is needed to register the device with vfio subsystem.

If i understand https://www.kernel.org/doc/Documentation/vfio.txt correctly (not an expert here), it looks like at the end qemu will just open the device group in `/dev/vfio/` and then ask with a system call for the right device via the PCI address. Since we are dealing with a device here, it should at the end get a major (probably 196) and minor number (sequential?) which we can then recreate via `mknod` inside the container. That might be a path for the future with less things mounted.

Reply:
@rmohr you raise an interesting question. Indeed libvirt handles the unbinding / registration with vfio itself. I think it will operate just fine if for some reason the group is already present.

Now, SR-IOV DP folks assume that the groups are pre-registered with vfio / uio before the first device is allocated. It's just that libvirt already does it for us, so we don't need to introduce a separate component / instructions to set vfio devices up. I think that's ok and simplifies setup.

> Do we have to give the device back if the container goes down dirty

I don't think so, unless we are going to allocate the same devices to both vfio and netlink users. (Not supported or planned for SR-IOV DP as of now.) Libvirt should detect the group is already registered and do nothing.

> That might be a path for the future with less things mounted.

I think with multiendpoints branch for SR-IOV DP, we can mount just /dev/vfio/vfio; and the appropriate /dev/vfio/<num> will be registered by DP itself. (We may need to create it, yes; for that, we can crawl the other /sys/* directories exposed. Not convenient but doable.)